### PR TITLE
all tests are created for book store

### PR DESCRIPTION
**Pull request by Abdul-Rehman2134**

### CODE REVIEW

File: tests/Feature/AuthorTest.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+class AuthorTest extends TestCase
+{
+    use RefreshDatabase;
+    /**
+     * A basic feature test example.
+     *
+     * @return void
+     */
+    public function test_author_list()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/AuthorTest.php:18]: change name to `test_list`

File: tests/Feature/AuthorTest.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        $author = Author::factory()->create();
+        $this->get(route('authors.index'))
+            ->assertSeeText('Authors')
+            ->assertSeeText('Name')
+            ->assertSeeText('Action')
+            ->assertSee($author->name);
+    }
+
+    // Author Add
+    public function test_author_add()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/AuthorTest.php:30]: change name to `test_create`

File: tests/Feature/AuthorTest.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    {
+        $this->login();
+        $author = Author::factory()->create();
+        $this->get(route('authors.index'))
+            ->assertSeeText('Add Author')
+            ->assertSeeText('Name');
+    }
+
+    // Author Create
+    public function test_author_create()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/AuthorTest.php:40]: change name to `test_store`

File: tests/Feature/AuthorTest.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        )->assertSessionDoesntHaveErrors();
+        $this->assertDatabaseHas((new Author())->getTable(),
+            [
+                'name' => $author->name
+            ]
+        );
+    }
+
+    // Author Edit
+    public function test_author_edit()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/AuthorTest.php:58]: `test_edit`

File: tests/Feature/AuthorTest.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    {
+        $this->login();
+        $author = Author::factory()->create();
+        $this->get(route('authors.index', $author->id))
+            ->assertSeeText('Edit Author')
+            ->assertSeeText('Name');
+    }
+
+    // Author Update
+    public function test_author_update()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/AuthorTest.php:68]: `test_update`

File: tests/Feature/AuthorTest.php [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        )->assertSessionDoesntHaveErrors();
+        $this->assertDatabaseHas((new Author())->getTable(),
+            [
+                'name' => $author->name
+            ]
+        );
+    }
+
+    // Author Delete
+    public function test_author_destroy()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/AuthorTest.php:86]: `test_destroy`

File: tests/Feature/BookTest.php [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+class BookTest extends TestCase
+{
+    use RefreshDatabase;
+    /**
+     * A basic feature test example.
+     *
+     * @return void
+     */
+    // Book Page
+    public function test_book()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/BookTest.php:21]: `test_index`

File: tests/Feature/BookTest.php [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        $this->get(route('books.index'))
+            ->assertSeeText('Books')
+            ->assertSee($book->name)
+            ->assertSee($book->author->name)
+            ->assertSee($book->category->name)
+            ->assertSee($book->price)
+            ->assertSeeText('QTY');
+    }
+    // Book Detail Page
+    public function test_book_detail()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/BookTest.php:35]: test_view

File: tests/Feature/BookTest.php [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+            ->assertSeeText('Action')
+            ->assertSeeText('Action')
+            ->assertSee($book->name)
+            ->assertSee($author->name)
+            ->assertSee($category->name)
+            ->assertSee($book->number_of_pages)
+            ->assertSee($book->price);
+    }
+    // Add Book
+    public function test_admin_add_book()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/BookTest.php:74]: `test_create`

File: tests/Feature/BookTest.php [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+        $author = Author::factory()->create();
+        $category = Category::factory()->create();
+        $book = Book::factory()->create(['author_id' => $author->id, 'category_id' => $category->id]);
+        $this->get(route('book.index', $book->id))
+            ->assertSeeText('Add Book')
+            ->assertSeeText('Name')
+            ->assertSeeText('Description');
+    }
+    // Book create
+    public function test_admin_book_store()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/BookTest.php:86]: test_store

File: tests/Feature/BookTest.php [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+    {
+        $this->login();
+        $author = Author::factory()->create();
+        $category = Category::factory()->create();
+        $book = Book::factory()->create(['author_id' => $author->id, 'category_id' => $category->id]);
```
Comment (AhmedAlBaloshi) [inline tests/Feature/BookTest.php:91]: remove book factory in this test

File: tests/Feature/BookTest.php [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+        $this->post(route('book.store'), [
+            'name' => $book->name,
```
Comment (AhmedAlBaloshi) [inline tests/Feature/BookTest.php:93]: add data manualy in post form

File: tests/Feature/BookTest.php [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+                'author_id' => $author->id,
+                'category_id' => $category->id,
+                'number_of_pages' => $book->number_of_pages,
+                'price' => $book->price,
+                'description' => $book->description
+            ]
+        );
+    }
+    // Edit Book
+    public function test_admin_edit_book()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/BookTest.php:115]: test_edit

File: tests/Feature/BookTest.php [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+    {
+        $this->login();
+        $author = Author::factory()->create();
+        $category = Category::factory()->create();
+        $book = Book::factory()->create(['author_id' => $author->id, 'category_id' => $category->id]);
+        $this->get(route('book.index', $book->id))
+            ->assertSeeText('Edit Book')
+            ->assertSeeText('Name')
```
Comment (AhmedAlBaloshi) [inline tests/Feature/BookTest.php:123]: add assertion  for book columns

File: tests/Feature/BookTest.php [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+            ->assertSeeText('Description');
+    }
+    // Book update
+    public function test_admin_book_update()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/BookTest.php:127]: test update

File: tests/Feature/BookTest.php [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+                'author_id' => $author->id,
+                'category_id' => $category->id,
+                'number_of_pages' => $book->number_of_pages,
+                'price' => $book->price,
+                'description' => $book->description
+            ]
+        );
+    }
+    // Book Delete
+    public function test_admin_book_delete()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/BookTest.php:156]: `test_destroy`

File: tests/Feature/BookTest.php [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+    {
+        $this->login();
+        $author = Author::factory()->create();
+        $category = Category::factory()->create();
+        $book = Book::factory()->create(['author_id' => $author->id, 'category_id' => $category->id]);
+        $this->post(route('book.destroy', $book->id))
+            ->assertSessionDoesntHaveErrors();
+        // ->assertRedirect();
```
Comment (AhmedAlBaloshi) [inline tests/Feature/BookTest.php:164]: remove useless commits

File: tests/Feature/CartTest.php [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+class CartTest extends TestCase
+{
+    use RefreshDatabase;
+    /**
+     * A basic feature test example.
+     *
+     * @return void
+     */
+    public function test_example()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/CartTest.php:20]: `test_index`

File: tests/Feature/CategoryTest.php [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+class CategoryTest extends TestCase
+{
+    use RefreshDatabase;
+    /**
+     * A basic feature test example.
+     *
+     * @return void
+     */
+    // Category List
+    public function test_category_list()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/CategoryTest.php:19]: `test_list`

File: tests/Feature/CategoryTest.php [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+        $category = Category::factory()->create();
+        $this->get(route('categories.index'))
+            ->assertSeeText('Categories')
+            ->assertSeeText('Name')
+            ->assertSeeText('Action')
+            ->assertSee($category->name);
+    }
+
+    // Category Add
+    public function test_category_add()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/CategoryTest.php:31]: `test_create`

File: tests/Feature/CategoryTest.php [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $this->login();
+        $category = Category::factory()->create();
+        $this->get(route('categories.index'))
+            ->assertSeeText('Add Category')
+            ->assertSeeText('Name');
+    }
+
+    // Category Create
+    public function test_category_create()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/CategoryTest.php:41]: `test_store`

File: tests/Feature/CategoryTest.php [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $this->login();
+        $category = Category::factory()->create();
```
Comment (AhmedAlBaloshi) [inline tests/Feature/CategoryTest.php:44]: In this case category factory is useless so remove it.

File: tests/Feature/CategoryTest.php [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+        $this->post(
+            route('categories.store', $category->id),
+            [
+                'name' => $category->name
```
Comment (AhmedAlBaloshi) [inline tests/Feature/CategoryTest.php:48]: add data manually

File: tests/Feature/CategoryTest.php [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+        )->assertSessionDoesntHaveErrors();
+        $this->assertDatabaseHas((new Category())->getTable(),
+            [
+                'name' => $category->name
+            ]
+        );
+    }
+
+    // Category Edit
+    public function test_category_edit()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/CategoryTest.php:59]: test_edit

File: tests/Feature/CategoryTest.php [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $this->login();
+        $category = Category::factory()->create();
+        $this->get(route('categories.index', $category->id))
+            ->assertSeeText('Edit Category')
+            ->assertSeeText('Name');
+    }
+
+    // Category Update
+    public function test_category_update()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/CategoryTest.php:69]: test_update

File: tests/Feature/CategoryTest.php [hunk 26]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $this->login();
+        $category = Category::factory()->create();
+        $this->post(
+            route('categories.update', $category->id),
+            [
+                'name' => $category->name
```
Comment (AhmedAlBaloshi) [inline tests/Feature/CategoryTest.php:76]: add data manually

File: tests/Feature/CategoryTest.php [hunk 27]
```diff
@@ -0,0 +1,93 @@
...
+        )->assertSessionDoesntHaveErrors();
+        $this->assertDatabaseHas((new Category())->getTable(),
+            [
+                'name' => $category->name
+            ]
+        );
+    }
+
+    // Category Delete
+    public function test_category_destroy()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/CategoryTest.php:87]: test_destroy

File: tests/Feature/ContactTest.php [hunk 28]
```diff
@@ -0,0 +1,24 @@
...
+
+class ContactTest extends TestCase
+{
+    use RefreshDatabase;
+    /**
+     * A basic feature test example.
+     *
+     * @return void
+     */
+    public function test_example()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/ContactTest.php:17]: test_index

File: tests/Feature/HomeTest.php [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+
+class HomeTest extends TestCase
+{
+    use RefreshDatabase;
+    /**
+     * A basic feature test example.
+     *
+     * @return void
+     */
+    public function test_example()
```
Comment (AhmedAlBaloshi) [inline tests/Feature/HomeTest.php:20]: test_index

**Review by AhmedAlBaloshi — CHANGES_REQUESTED** [review]

File: tests/Feature/AuthorTest.php (hunk 1 above)
Comment (Abdul-Rehman2134) [reply tests/Feature/AuthorTest.php:18]: done

**Review by AhmedAlBaloshi — APPROVED** [review]

**Pull request merged**